README.md
# AC IP
This is a little helper for IP and network operations.

This is currently just a preliminary version with limited documentation.

## Usage

```
const acts = require('ac-ip')
```

### Determine IP
Determines the IP from the request object
```
const ip = acts.determineIP(req) 
// -> 1.2.3.4
```
Use X-AdmiralCloud-Test "true" to overwrite the real IP with params.ip from request object.


### IP to privacy
Makes the IP GDRP ready by removing the last two octecs of the IP(s)
```
const privacyIP = acts.ipsToPrivacy(['1.2.3.4', ...])
// -> ['1.2.x.x', ...]
```

### Check if IP is private
```
const isPrivate = acts.isPrivate('1.2.3.4')
// -> false

const isPrivate = acts.isPrivate('127.0.0.1')
// -> true
```

# Error codes
All errors have a message, but messages can change. Therefor all error messages now also have an error code:

| Code | Message |
|---|---|
| 9000 | acip_determineIP_noIPDetected |
| 9001 | acip_checkCIDR_listIsEmpty |
| 9002 | acip_checkCIDR_ipNotInCIDRrange |
| 9003 | acip_checkCIDR_cidrIsNotValid |
| 9004 | acip_checkCIDR_thisIsNoCIDR |
| 9005 | acip_checkCIDR_maskInvalid |
| 9006 | acip_checkCIDR_invalid |
| 9007 | acip_checkCIDR_maskInvalid |





## Links
- [Website](https://www.admiralcloud.com/)
- [Twitter (@admiralcloud)](https://twitter.com/admiralcloud)
- [Facebook](https://www.facebook.com/MediaAssetManagement/)

## License
[MIT License](https://opensource.org/licenses/MIT) Copyright © 2009-present, AdmiralCloud, Mark Poepping

CHANGELOG.md
<a name="1.3.5"></a>

## [1.3.5](https://github.com/mmpro/ac-ip/compare/v1.3.4..v1.3.5) (2021-05-08 07:22:35)


### Bug Fix

* **App:** Force version bump | MP | [0cae9b46eb613b86bd6471fe8940fd9c9c4f1a30](https://github.com/mmpro/ac-ip/commit/0cae9b46eb613b86bd6471fe8940fd9c9c4f1a30)    
Force version bump
### Tests

* **Misc:** Added test | MP | [9093f7b0bb5c4cf58fb0ce9f41fa708a3f406220](https://github.com/mmpro/ac-ip/commit/9093f7b0bb5c4cf58fb0ce9f41fa708a3f406220)    
Test coverage is now ~99%
### Chores

* **App:** Updated gitignore | MP | [29748cfe061af03bd162e39f388b1e4537b60c6f](https://github.com/mmpro/ac-ip/commit/29748cfe061af03bd162e39f388b1e4537b60c6f)    
Updated gitignore
### Chores

* **Misc:** Updated packages | MP | [f4ace7c58b7c1d60b01c1c6634cf816b4bceadf8](https://github.com/mmpro/ac-ip/commit/f4ace7c58b7c1d60b01c1c6634cf816b4bceadf8)    
Updated packages
<a name="1.3.4"></a>

## [1.3.4](https://github.com/mmpro/ac-ip/compare/v1.3.3..v1.3.4) (2020-05-12 08:50:47)


### Bug Fix

* **App:** Add error codes | MP | [3e70c7818554c3837ce4ebeec768fe95352212b4](https://github.com/mmpro/ac-ip/commit/3e70c7818554c3837ce4ebeec768fe95352212b4)    
Add error codes in addition to error message
### Tests

* **App:** Fixed test | MP | [cf5a8d8399e597ad4d8e1a0e6b1dde43255941f2](https://github.com/mmpro/ac-ip/commit/cf5a8d8399e597ad4d8e1a0e6b1dde43255941f2)    
Add code to error response
### Chores

* **App:** Updated packages | MP | [83c937eeed2eab9bba89974277b87fd52ef79885](https://github.com/mmpro/ac-ip/commit/83c937eeed2eab9bba89974277b87fd52ef79885)    
Updated packages
* **undefined:** Updated packages | MP | [18845753d849c612113a9a6b0a74bd5454db0da8](https://github.com/mmpro/ac-ip/commit/18845753d849c612113a9a6b0a74bd5454db0da8)    
Updated packages
<a name="1.3.3"></a>

## [1.3.3](https://github.com/mmpro/ac-ip/compare/v1.3.2..v1.3.3) (2020-04-03 10:14:20)


### Bug Fix

* **App:** Fallback to development environment if none set | MP | [83c11ce93938ff00e2e70e0972bd9746fe032abd](https://github.com/mmpro/ac-ip/commit/83c11ce93938ff00e2e70e0972bd9746fe032abd)    
If process has not set any environment via NODE_ENV, then fallback to development
<a name="1.3.2"></a>

## [1.3.2](https://github.com/mmpro/ac-ip/compare/v1.3.1..v1.3.2) (2020-04-01 09:10:16)


### Bug Fix

* **App:** Use statusCode 420 in message if ip is not in range | MP | [995b22dd92a4914fc6272a5e24af588e3c328982](https://github.com/mmpro/ac-ip/commit/995b22dd92a4914fc6272a5e24af588e3c328982)    
IP no in range is not typcical error, but more like a preconditon fail. That's why we now use statusCode 420 in addition to the error message. This way you can decide whether you want to treat it as error or as warning.
<a name="1.3.1"></a>

## [1.3.1](https://github.com/mmpro/ac-ip/compare/v1.3.0..v1.3.1) (2020-03-29 14:04:57)


### Bug Fix

* **App:** Prepare repository for AC semantic release  | MP | [30dadc928a5ed428d73b01ea0efe1afc1dffc94c](https://github.com/mmpro/ac-ip/commit/30dadc928a5ed428d73b01ea0efe1afc1dffc94c)    
Cleaned up repository and use ac-semantic-release
<a name="1.3.0"></a>
# [1.3.0](https://github.com/mmpro/ac-ip/compare/v1.2.0...v1.3.0) (2020-03-18 11:05)


### Features

* **Misc:** checkCIDR now returns the matching CIDR instead of true | MP ([8a34f26c328d666ae803f5bef3b6aa5f14677051](https://github.com/mmpro/ac-ip/commit/8a34f26c328d666ae803f5bef3b6aa5f14677051))    
  checkCIDR now returns the matching CIDR instead of true



<a name="1.2.0"></a>
# [1.2.0](https://github.com/mmpro/ac-ip/compare/v1.1.4...v1.2.0) (2019-11-08 10:32)


### Features

* **Misc:** CHeck if IP is private | MP ([a83311517acfa828c1ca6f2fe9dc42a0d986a955](https://github.com/mmpro/ac-ip/commit/a83311517acfa828c1ca6f2fe9dc42a0d986a955))    
  New function "isPrivate" to check if a given IP is public or private



<a name="1.1.4"></a>
## [1.1.4](https://github.com/mmpro/ac-ip/compare/v1.1.3...v1.1.4) (2019-10-06 12:53)


### Bug Fixes

* **Misc:** Send IP as parameter if not in production mode | MP ([2a5725c](https://github.com/mmpro/ac-ip/commit/2a5725c))    
  For easier debugging it is now possible to just send the IP address as query parameter.



<a name="1.1.3"></a>
## [1.1.3](https://github.com/mmpro/ac-ip/compare/v1.1.2...v1.1.3) (2019-10-06 09:15)


### Bug Fixes

* **Misc:** Updated packages / Security fix | MP ([8e4c21c](https://github.com/mmpro/ac-ip/commit/8e4c21c))    
  Updated packages / Security fix



<a name="1.1.2"></a>
## [1.1.2](https://github.com/mmpro/ac-ip/compare/v1.1.1...v1.1.2) (2019-07-24 19:28)


### Bug Fixes

* **Misc:** Updated packages | MP ([8cec625](https://github.com/mmpro/ac-ip/commit/8cec625))    
  Updated packages



<a name="1.1.1"></a>
## [1.1.1](https://github.com/mmpro/ac-ip/compare/v1.1.0...v1.1.1) (2019-07-03 15:56)


### Bug Fixes

* **Misc:** Return unique array for ipsToPrivacy | MP ([c51318f](https://github.com/mmpro/ac-ip/commit/c51318f))    
  Make sure to group ips



<a name="1.1.0"></a>
# [1.1.0](https://github.com/mmpro/ac-ip/compare/v1.0.4...v1.1.0) (2019-07-03 15:20)


### Features

* **Misc:** New functions to make IPs GDPR ready and to check IPs against a list of IPs | MP ([40c2e24](https://github.com/mmpro/ac-ip/commit/40c2e24))    
  New functions to make IPs GDPR ready and to check IPs against a list of IPs



<a name="1.0.4"></a>
## [1.0.4](https://github.com/mmpro/ac-ip/compare/v1.0.3...v1.0.4) (2019-06-09 17:16)


### Bug Fixes

* **Misc:** Do not user req.allParams | MP ([57e5e4e](https://github.com/mmpro/ac-ip/commit/57e5e4e))    
  req.allParams is not available for non-sails APIs



<a name="1.0.3"></a>
## [1.0.3](https://github.com/mmpro/ac-ip/compare/v1.0.2...v1.0.3) (2019-04-09 16:22)


### Bug Fixes

* **Misc:** Allow custom IP with X-AdmiralCloud-Test header | MP ([89cc08c](https://github.com/mmpro/ac-ip/commit/89cc08c))    
  If you want to set a custom IP for testing or development, use X-AdmiralCloud-Test header as true

test/test.js
const expect = require('expect')
const acip = require('../index')
const _ = require('lodash')

describe('Determine IP from request object', () => {
  it('Test IP', done => {
    let req = { 
      ip: '8.8.8.8'
    }
    let test = acip.determineIP(req)
    expect(test).toEqual('8.8.8.8')
    return done()  
  })

  it('Test Proxy IP - with x-real-ip', done => {
    let req = { 
      ip: '8.8.8.8',
      headers: {
        'x-real-ip': '4.4.4.4'
      }
    }
    let test = acip.determineIP(req)
    expect(test).toEqual('4.4.4.4')
    return done()  
  })

  it('Test Proxy IP - with x-forwarded-for', done => {
    let req = { 
      ip: '8.8.8.8',
      headers: {
        'x-forwarded-for': '1.1.1.1'
      }
    }
    let test = acip.determineIP(req)
    expect(test).toEqual('1.1.1.1')
    return done()  
  })

  it('Test Proxy IP - with both headers - prefer x-real-ip', done => {
    let req = { 
      ip: '8.8.8.8',
      headers: {
        'x-real-ip': '4.4.4.4',
        'x-forwarded-for': '1.1.1.1'
      }
    }
    let test = acip.determineIP(req)
    expect(test).toEqual('4.4.4.4')
    return done()  
  })

  it('Test multiple Proxy IP - with x-forwarded-for', done => {
    let req = { 
      ip: '8.8.8.8',
      headers: {
        'x-forwarded-for': '1.1.1.1, 4.4.4.4, 1.2.3.4'
      }
    }
    let test = acip.determineIP(req)
    expect(test).toEqual('1.1.1.1')
    return done()  
  })

  it('Test set ip in test mode', done => {
    let req = { 
      query: {
        ip: '5.4.1.2'
      },
      ip: '8.8.8.8',
      headers: {
        'x-real-ip': '4.4.4.4',
        'x-forwarded-for': '1.1.1.1'
      }
    }
    let test = acip.determineIP(req)
    expect(test).toEqual('5.4.1.2')
    return done()  
  })

})

describe('Testing CIDR', function () {
  const validCIDRs = [{ cidr: '8.8.0.0/16' }]
  const invalidCIDRs = [{ cidr: '8.8.0.0' }]
  const validv6CIDRs = [{ cidr: '2001:4d20::/32', type: 'ipv6' }]

  let test1 = '85.182.224.128'
  let test2 = [{ 'cidr': '85.182.224.128/26' }, { 'cidr': '62.96.36.146/32' }, { 'cidr': '62.84.220.138/32' }]

  it('Check if CIDR is valid', function(done) {
    let test = acip.checkCIDR({ cidr: validCIDRs })
    expect(test).toBeUndefined()
    return done()
  })

  it('Check if CIDR IPv6 is valid', function(done) {
    let test = acip.checkCIDR({ cidr: validv6CIDRs })
    expect(test).toBeUndefined()
    return done()
  })

  it('Check if CIDR is invalid', function(done) {
    let test = acip.checkCIDR({ cidr: invalidCIDRs })
    expect(test).toEqual({ 'additionalInfo': { 'cidr': '8.8.0.0', 'type': 'ipv4' }, code: 9004, 'message': 'acip_checkCIDR_thisIsNoCIDR' })
    return done()
  })

  it('Check if IP is in CIDR ', function(done) {
    let test = acip.checkCIDR({ ip: test1, cidr: test2 })
    expect(test).toBe(null)
    return done()
  })

  it('Check if IP is in CIDR using callback ', function(done) {
    acip.checkCIDR({ ip: test1, cidr: test2 }, (err, result) => {
      if (err) return done(err)
      expect(result).toEqual(_.get(_.first(test2), 'cidr'))
      return done()
    })
  })

  it('Check CIDR without parameter CIDR - fail', function(done) {
    let test = acip.checkCIDR({ ip: test1 })
    expect(test.code).toEqual(9001)
    expect(test.message).toEqual("acip_checkCIDR_listIsEmpty")
    return done()
  })

  it('Check CIDR without invalid mask - fail', function(done) {
    let test = acip.checkCIDR({ cidr: [{ cidr: '85.182.224.128/333' }] })
    expect(test.code).toEqual(9005)
    expect(test.message).toEqual("acip_checkCIDR_maskInvalid")
    return done()
  })

  it('Check CIDR without invalid mask ipv6 - fail', function(done) {
    let test = acip.checkCIDR({ cidr: [{ cidr: '2001:4d20::/323', type: 'ipv6' }] })
    expect(test.code).toEqual(9007)
    expect(test.message).toEqual("acip_checkCIDR_maskInvalid")
    return done()
  })

  it('Check CIDR with invalid cidr - fail', function(done) {
    let test = acip.checkCIDR({ cidr: [{ cidr: '85.182.224.128333/8' }] })
    expect(test.code).toEqual(9006)
    expect(test.message).toEqual("acip_checkCIDR_invalid")
    return done()
  })

  /* IP package has possible bug: https://github.com/indutny/node-ip/issues/59
  it('Check CIDR with invalid cidr ipv6 - fail', function(done) {
    let test = acip.checkCIDR({ cidr: [{ cidr: 'a356/64', type: 'ipv6' }] })
    expect(test.code).toEqual(9006)
    expect(test.message).toEqual("acip_checkCIDR_invalid")
    return done()
  })
  */

  it('Return IP block from CIDR', function(done) {
    let result = acip.ipsFromCIDR({ cidr: '8.8.8.8/31' })
    expect(result).toEqual(["8.8.8.8", "8.8.8.9"])
    return done()
  })
})

describe('IPs to privacy', () => {
  const ips = ['8.8.8.8', '4.4.4.4']

  it('Check that IPs are masked properly', (done) => {
    let test = acip.ipsToPrivacy(ips)
    expect(test).toEqual(['8.8.x.x', '4.4.x.x'])
    return done()
  })
})

describe('IP in IP list', () => {
  const ip = '4.4.x.x'
  const ipFail = '1.1.x.x'
  const ips = ['8.8.x.x', '4.4.x.x']

  it('Check that IP is in list', (done) => {
    let test = acip.ipInIPList({ ips, ip })
    expect(test).toEqual(true)
    return done()
  })

  it('Check that IP is not in list', (done) => {
    let test = acip.ipInIPList({ ips, ipFail })
    expect(test).toEqual(false)
    return done()
  })
})

describe('Is IP private', () => {
  const privateIPv4 = '127.0.0.1'
  const privateIPv6 = '::ffff:172.31.7.78'
  const publicIPv4 = '8.8.8.8'
  const publicIPv6 = '2001:4860:4860::8888'

  it('Check private IPv4 address', done => {
    let test = acip.isPrivate(privateIPv4)
    expect(test).toEqual(true)
    return done()
  })

  it('Check private IPv6 address', done => {
    let test = acip.isPrivate(privateIPv6)
    expect(test).toEqual(true)
    return done()
  })

  it('Check public IPv4 address', done => {
    let test = acip.isPrivate(publicIPv4)
    expect(test).toEqual(false)
    return done()
  })

  it('Check public IPv6 address', done => {
    let test = acip.isPrivate(publicIPv6)
    expect(test).toEqual(false)
    return done()
  })

})


.czConfig.js
module.exports = {
  repository: {
    baseUrl: 'https://github.com/',
    repoUrl: 'mmpro/REPOSITORY/'
  },
  types: [
    {value: 'feat',     name: 'feat:     A new feature'},
    {value: 'fix',      name: 'fix:      A bug fix'},
    {value: 'docs',     name: 'docs:     Documentation only changes'},
    {value: 'style',    name: 'style:    Changes that do not affect the meaning of the code\n            (white-space, formatting, missing semi-colons, etc)'},
    {value: 'refactor', name: 'refactor: A code change that neither fixes a bug nor adds a feature'},
    {value: 'perf',     name: 'perf:     A code change that improves performance'},
    {value: 'test',     name: 'test:     Adding missing tests'},
    {value: 'chore',    name: 'chore:    Changes to the build process or auxiliary tools\n            and libraries such as documentation generation'},
    {value: 'revert',   name: 'revert:   Revert to a commit'},
    {value: 'WIP',      name: 'WIP:      Work in progress'}
  ],
  scopes: [
    {name: 'Misc' }
  ],
  allowCustomScopes: true,
  allowBreakingChanges: ['feat', 'fix']
};


index.js
const _ = require('lodash')
const ipPackage = require('ip')

const acip = () => {
  /**
   * Send an express-like request object into the function, the IP address of the request is returned.
   *
   * If you send req.debugMode = true, you can manually overwrite the IP with a payload parameter "ip"
   *
   * @param req object Express-like request object
   */
  const determineIP = (req) => {
    const proxyIP = _.get(req, 'headers.x-real-ip') || _.get(req, 'headers.x-forwarded-for')
    let ip = proxyIP || _.get(req, 'ip')
    // allow "overwriting" IP for local testing, but not in production, send X-AdmiralCloud-Header "true"
    if (_.has(req, 'query.ip') && _.indexOf(['development', 'test'], _.get(process, 'env.NODE_ENV', 'development')) > -1) {
      ip = _.get(req, 'query.ip')
    }

    // LEGACY - REMOVE 2019-06-30
    if (req.debugMode && _.has(req, 'query.ip')) ip = _.get(req, 'query.ip')

    if (!ip) return { code: 9000, message: 'acip_determineIP_noIPDetected' }
    // x forwarded for can be a comma or space separated list - z.b. 192.168.24.73, 198.135.124.15
    // X-Forwarded-For: client1, proxy1, proxy2 -> but client1 can be a private IP address
    if (ip.indexOf(',') > -1) {
      // check until we've found a non-private IP address
      let finalIP
      _.some(ip.split(','), (ipToCheck) => {
        if (!ipPackage.isPrivate(_.trim(ipToCheck))) {
          finalIP = _.trim(ipToCheck)
          return true
        }
      })
      ip = finalIP
    }
    return ip
  }

  /**
   * Returns the list of IP adresses for a given CIDR block
   * @param params.cidr STRING -> valid cidr block - e.g. 192.168.1.134/26
   * ipsFromCIDR({ cidr: '8.8.8.8/31' })
   * Returns ['8.8.8.8', '8.8.8.9']
   */
  const ipsFromCIDR = function(params) {
    const cidr = params.cidr
    const regex = /(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})\/(\d{1,2})/
    let match = cidr.match(regex)
    const suffix = _.get(match, '[5]')
    if (!suffix || parseInt(suffix) < 24) return []
    // check that cidr is valid (structural check)
    let range = ipPackage.cidrSubnet(cidr)
    let ips = []
    let start = ipPackage.toLong(_.get(range, 'firstAddress')) // 2130706433
    let end = ipPackage.toLong(_.get(range, 'lastAddress')) // 2130706433
    for (let x = start; x <= end; x += 1) {
      ips.push(ipPackage.fromLong(x))
    }
    return ips
  }

  /**
   * Checks an array (list) of CIDR for validity
   *
   * @param params.cidr array of objects containing keys "cidr" and optional "type". If no type => ipv4
   * @param params.ip string ip address to check against the given cidrs -> return true if matching
   * @param params.noMatchAllowed BOOL Only in combination with param.ip - if true, no error if returned if there is no match for the IP in given CIDR
   *
   * @param params.cb OPT (err, match) -> match is the matching CIDR if ip is given
   *
   * [{
    "cidr": "2001:280::/32",
    "type"; "ipv6"
   },...]
   *
   */

  const checkCIDR = (params, cb) => {
    // 1 check if array
    const cidr = _.isArray(params.cidr) && params.cidr.length > 0 && params.cidr
    if (!cidr) {
      if (cb) return cb({ code: 9001, message: 'acip_checkCIDR_listIsEmpty' })
      return { code: 9001, message: 'acip_checkCIDR_listIsEmpty' }
    }

    if (params.ip) {
      let error = { code: 9002, message: 'acip_checkCIDR_ipNotInCIDRrange', statusCode: 420 }
      if (params.noMatchAllowed) error = null

      // check if IP is a match for any of the given CIDR
      let match
      _.some(cidr, (c) => {
        if (ipPackage.cidrSubnet(c.cidr).contains(params.ip)) {
          error = null
          match = c.cidr 
          return true
        }
      })
      if (cb) return cb(error, match)
      return error
    }
    else {
      // check if all cidrs are valid
      let error
      _.some(cidr, c => {
        if (!_.isString(c.cidr)) error = { code: 9003, message: 'acip_checkCIDR_cidrIsNotValid' }
        else if (c.cidr.indexOf('/') < 0) error = { code: 9004, message: 'acip_checkCIDR_thisIsNoCIDR' }
        else if (!c.type || c.type === 'ipv4') {
          // check mask (max is 32)
          let mask = _.last(_.split(_.get(c, 'cidr', ''), '/'))
          if (mask > 32) {
            error = { code: 9005, message: 'acip_checkCIDR_maskInvalid' }
          }
          else if (!ipPackage.isV4Format(ipPackage.cidr(c.cidr))) {
            error = { code: 9006, message: 'acip_checkCIDR_invalid' }
          }
        }
        else if (c.type === 'ipv6') {
          // check mask (max is 128)
          let mask = _.last(_.split(_.get(c, 'cidr', ''), '/'))
          if (mask > 128) {
            error = { code: 9007, message: 'acip_checkCIDR_maskInvalid' }
          }
          else if (!ipPackage.isV6Format(ipPackage.cidr(c.cidr))) {
            error = { code: 9006, message: 'acip_checkCIDR_invalid' }
          }
        }

        if (error) {
          _.merge(error, { additionalInfo: { cidr: c.cidr, type: _.get(c, 'type', 'ipv4') } })
        }
        return error
      })
      if (cb) return cb(error)
      return error
    }
  }

  /**
   * Ingests a list of IPv4 addresses and returns them with the last 2 octets removed/replaced with xxx
   * @param {*} params.ips
   */
  const ipsToPrivacy = (ips) => {
    const regex = /(\d{1,3}\.\d{1,3}\.)(\d{1,3}\.\d{1,3})/
    ips = _.map(_.compact(_.uniq(ips)), ip => {
      return ip.replace(regex, (match, p1) => {
        return p1 + 'x.x'
      })
    })
    return _.uniq(ips)
  }

  /**
   * Checks is an IP is in an IP list
   * @param params.ip STRING ip IP address to check
   * @param params.ips ARRAY list of ips
   */
  const ipInIPList = (params) => {
    const ip = _.get(params, 'ip')
    const ips = _.uniq(_.get(params, 'ips'))
    return _.indexOf(ips, ip) >= 0
  }

  const isPrivate = (ip) => {
    return ipPackage.isPrivate(_.trim(ip))
  }

  return {
    determineIP,
    checkCIDR,
    ipsFromCIDR,
    ipsToPrivacy,
    ipInIPList,
    isPrivate
  }
}

module.exports = acip()